### DISCUSSION TITLE
Make the terminal feel more like the elementary Terminal

### DISCUSSION THREAD

**Discussion by hanaral** [elementary/code#938] [Ideas]
## Problem

I personally feel that it is annoying to have to switch between 2 different brain modes when I'm either using the Terminal app, or Code's terminal. It doesn't copy and paste the same way and doesn't have at least a toggle for that behaviour. The terminal widget also doesn't look too great, and apart from the colour scheme looks to be the VTE widget crammed in a box.

## Proposal

I think that the actual behavour of the terminal should be the same as the Terminal app
It should:
- Have a toggle in the settings or havea small top bar with the same copy+paste widget Terminal has.
- Not have an odd box around the widget
- Potentially just be a feature and not a plugin
- Maybe even be transparent like Terminal and be able to see your code beneath

And yes, I could just use Terminal but that doesn't integrate with a code editing workflow.

**Comment by danirabbit** [comment]
Converted to a discussion since this isn’t really an actionable issue report but a kind of collection of ideas

My personal view is that trying to make Code’s Terminal plugin behave exactly like the Terminal app is probably not a great use of resources and would end up in either a lot of code duplication or complicating the Terminal app a lot to try to make it into some kind of library. It just doesn’t sound like a great thing to pursue.

I think it would probably be more scalable to actually remove the Terminal plugin and instead build in better git and meson features which probably accomplishes what you’re actually using the plugin for anyways

**Reply by hanaral** [reply]
Ok even better, I literally just opened an feature request for compiling controls and elementary-app-development-stuff, you can close that too if it'll be better to have the entire thing here

**Comment by hanaral** [comment]
_(I'm aware a lot of these can only be aims and not feasable reource-wise atm)_
Some ideas for developer-focused features:
![Screenshot from 2021-02-06 00-56-12](https://user-images.githubusercontent.com/58219504/107103668-5d5edf80-6816-11eb-9520-dfa89e1f1f0e.png)
A lot of these could just be 'sane defaults', like having a dedicated sidebar with tabs for different things like Git actions (such as the ability to log in with a GH account)

**Comment by cassidyjames** [comment]
Another idea is to retain some sort of "Terminal" functionality by having a button that opens the current project in the Terminal app. It would be pretty simple implementation-wise, but might serve that want for "I want a button that opens a Terminal in this project", and would also hammer home the idea that our IDE is Code + Terminal, not just one or the other.

This could be in addition to better Git, Meson, Flatpak, etc. support directly in Code itself.

**Reply by hanaral** [reply]
Ok so after some thinking, I agree with @cassidyjames. Are any of you willing to close this? cuz I can't

**Reply by jeremypw** [reply]
I always open a project in Terminal using the project folder context menu - it only takes two clicks - because a lot of the workflow cannot be done in Code.  Over time I expect more and more to be doable in Code but for many things it is convenient to have a separate window for output.  It would be even more convenient to have a button to do this but not sure where to put it as there is already an action icon on the project folder row.  I suppose, if it is on the headerba, it would have to work out which is the active project from the active document?

**Comment by jeremypw** [comment]
Making the terminal more like the native Terminal app will be more challenge in the Flatpak version because of the sandboxing.
We should probably move towards making the terminal pane redundant as far as possible by providing more convenient UI ways of achieving the same thing.  

Some progress has been made by providing ways of cloning repos, opening and creating branches for example.  Other branch operations are planned.

**Reply by zeebok** [reply]
Is there any known line that makes Code good enough to remove the terminal? You mention some Git operations but what else comes to mind? To me it would be nice to have a configurable button to build/run using commands and a panel that catches console output for that button. Pretty big feature but personally would cover the last reason to have the terminal besides the Git stuff

**Reply by jeremypw** [reply]
In principle all necessary commands for code development could be made accessible in the UI and the terminal replaced with a simple pane showing the output from the command.  In many cases typing a command is much quicker than going through a UI though (not all commands can have their own button/shortcut).  I think we would have to abandon a Terminal clone in the Flatpak but could probably still maintain some less accessible way of running a command on the host for use cases not covered by UI.
